### PR TITLE
Implement Subcategories

### PR DESCRIPTION
**Pull request by patmbolger**
## Changes proposed in this pull request:
1. See PT story

*NOTE*: ~~I've added uniqueness validation to BusinessCategory name.  We have been implicitly assuming this.~~ << I removed this as too many tests broke.  Will add separate story.

## Screenshots (Optional):

---
## Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: app/models/company.rb [hunk 1]
```diff
@@ -164,11 +164,22 @@ def any_visible_addresses?
   end
 
 
-  def categories_names
+  def categories_names(include_subcategories)
```
Comment (weedySeaDragon) [inline app/models/company.rb]: should set a default value for `include_subcategories`  (improves readability/comprehension in addition to adding type safety)

File: app/models/shf_application.rb [hunk 2]
```diff
@@ -50,6 +50,49 @@ class ShfApplication < ApplicationRecord
   CAN_EDIT_STATES = [:new, :waiting_for_applicant]
 
 
+  def business_subcategories(business_category)
```
Comment (weedySeaDragon) [inline app/models/shf_application.rb]: instance methods should go after class methods  (Ruby Style Guide: https://rubystyle.guide/#classes-modules_

File: app/models/shf_application.rb [hunk 3]
```diff
@@ -50,6 +50,49 @@ class ShfApplication < ApplicationRecord
...
+    return nil unless business_category.is_root?
+    return nil unless business_categories.include?(business_category)
+
+    subcategories = []
+
+    business_category_ids =  self.business_categories.map(&:id)
+
+    BusinessCategory.children_of(business_category.id).order(:name).each do |subcategory|
+      if business_category_ids.include?(subcategory.id)
```
Comment (weedySeaDragon) [inline app/models/shf_application.rb]: change this to 1 line: `subcategories << subcategory if business_category_ids.include?(subcategory.id)`

File: app/models/shf_application.rb [hunk 4]
```diff
@@ -50,6 +50,49 @@ class ShfApplication < ApplicationRecord
   CAN_EDIT_STATES = [:new, :waiting_for_applicant]
 
 
+  def business_subcategories(business_category)
+    return nil unless business_category.is_root?
+    return nil unless business_categories.include?(business_category)
+
+    subcategories = []
+
+    business_category_ids =  self.business_categories.map(&:id)
```
Comment (weedySeaDragon) [inline app/models/shf_application.rb]: is `self.` needed?

File: app/models/shf_application.rb [hunk 5]
```diff
@@ -50,6 +50,49 @@ class ShfApplication < ApplicationRecord
...
+      end
+    end
+
+    subcategories
+  end
+
+  def set_business_subcategories(business_category, subcategories)
+    # subcategories can be 1) an active record relation, 2) an array of
+    # BusinessCategory records, or 3) a single BusinessCategory record.
+
```
Comment (weedySeaDragon) [inline app/models/shf_application.rb]: Where is this used such that `subcategories` is a single BusinessCategory record?
If it is used with that, then it should call a different method that converts the single record to an array and then calls this one.  (Checking a class is a smell.)  Also:  need to include that case/context in the RSpec for this method.

File: app/views/payor/_term_status.html.haml [hunk 6]
```diff
@@ -30,7 +30,8 @@
   .status-details
     .row.mt-2.justify-content-between
       .flex.status
-        %span.label= status_label.html_safe
+        = succeed ':' do
```
Comment (weedySeaDragon) [inline app/views/payor/_term_status.html.haml:33]: I did not know about the `succeed` HAML helper.  I'm sure there are a bunch of views where we (I) should use this!

File: app/views/shf_applications/_business_category_display_row.html.haml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+-# Local vars -
+  shf_application: a member's SHF application
+  business_category: the BusinessCategory instance
+
+%tr.business_category.category-display-row{ id: "subcategories-display-row-#{business_category.id}" }
```
Comment (weedySeaDragon) [inline app/views/shf_applications/_business_category_display_row.html.haml:5]: `.business_category` should use a dash instead of an underscore

File: app/views/shf_applications/_business_category_display_row.html.haml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+  %td.name= business_category.name
+
+  %td.name
+    - subcategories = shf_application.business_subcategories(business_category)
+    - if subcategories
+      = subcategories.map(&:name).join(', ')
+    - else
+      None
```
Comment (weedySeaDragon) [inline app/views/shf_applications/_business_category_display_row.html.haml]: shouldn't this be `t('none')` ?

File: app/views/shf_applications/_business_category_edit_row.html.haml [hunk 9]
```diff
@@ -0,0 +1,49 @@
+-# Local vars -
+  shf_application: a member's SHF application
+  business_category: the BusinessCategory instance
+  subcategories: subcategories currently assigned to shf_application
+  language: language identifier
+
+%tr.business_category.category-edit-row{ id: "subcategories-edit-row-#{business_category.id}" }
```
Comment (weedySeaDragon) [inline app/views/shf_applications/_business_category_edit_row.html.haml:7]: `.business_category` should use a dash instead of an underscore

File: features/conditions/company_incomplete_alerts.feature [hunk 10]
```diff
@@ -59,9 +59,9 @@ Feature: Alerts sent if Company info is incomplete (a Condition response)
       | memb03_paid_jan10@blank-co-name.se | 2018-1-10  | 2019-01-09  | member_fee   | betald | none    |                |
       | memb03_paid_jan10@blank-co-name.se | 2018-1-10  | 2019-01-09  | branding_fee | betald | none    | 2120000142     |
       | memb04_paid_jan15@blank-co-name.se | 2018-1-15  | 2019-01-14  | member_fee   | betald | none    |                |
-      | memb20@complete_company.se         | 2018-1-1   | 2018-12-31  | member_fee   | betald | none    |                |
-      | memb20@complete_company.se         | 2018-1-1   | 2018-12-31  | branding_fee | betald | none    | 2664181183     |
-      | memb21@complete_company.se         | 2018-1-1   | 2018-12-31  | member_fee   | betald | none    |                |
+      | memb20@complete-company.se         | 2018-1-1   | 2018-12-31  | member_fee   | betald | none    |                |
```
Comment (weedySeaDragon) [inline features/conditions/company_incomplete_alerts.feature:62]: Just curious about why this change? (underscore in the email changed to a dash)

File: features/search_companies.feature [hunk 11]
```diff
@@ -17,11 +18,11 @@ Background:
...
-    | name         |
-    | Groomer      |
-    | Psychologist |
-    | Trainer      |
-    | Walker       |
+    | name         | subcategories |
+    | Groomer      | Groomer Subcat1, Groomer Subcat2 |
+    | Psychologist | Psycho Subcat1                   |
+    | Trainer      | Trainer Subcat1, Trainer Subcat2 |
+    | Walker       | Walker Subcat1, Walker Subcat2   |
```
Comment (weedySeaDragon) [inline features/search_companies.feature:25]: Should include a category with no subcategories so that it is clear to that having no subcategories is valid.

File: features/search_companies.feature [hunk 12]
```diff
@@ -111,6 +112,29 @@ Scenario: Search by category
...
 
+@selenium @time_adjust
+Scenario: Search by subcategory
+  Given I am Logged out
+  And I am on the "landing" page
+  And I should see "Barky Boys" in the list of companies
+  And I should see "HappyMutts" in the list of companies
+  And I should see "Dogs R Us" in the list of companies
+  And I should see "We Luv Dogs" in the list of companies
+  Then I select "Groomer Subcat1" in select list t("activerecord.models.business_category.one")
```
Comment (weedySeaDragon) [inline features/search_companies.feature:123]: "Then" should be "When"  (or "And", and the next step should be "When")   'Given-When-Then'

File: features/search_companies.feature [hunk 13]
```diff
@@ -111,6 +112,29 @@ Scenario: Search by category
...
+  And I click on t("search")
+  And I should see "Barky Boys" in the list of companies
```
Comment (weedySeaDragon) [inline features/search_companies.feature:125]: "And" should be "Then"

File: features/search_companies.feature [hunk 14]
```diff
@@ -111,6 +112,29 @@ Scenario: Search by category
...
+  And I should see "DogCo_01" in the list of companies
+  And I should not see "We Luv Dogs" in the list of companies
+  And I should not see "HappyMutts" in the list of companies
+  And I should not see "Dogs R Us" in the list of companies
+  And I should see "Trainer" in the list of companies
+  And I should see "Groomer" in the list of companies
+  And I should not see "Psychologist" in the list of companies
+  And I should not see "Walker" in the list of companies
+  Then I select "Trainer Subcat2" in select list t("activerecord.models.business_category.one")
```
Comment (weedySeaDragon) [inline features/search_companies.feature:134]: "Then" should be "When"

File: features/search_companies.feature [hunk 15]
```diff
@@ -111,6 +112,29 @@ Scenario: Search by category
...
+  And I click on t("search")
+  And I should see "Dogs R Us" in the list of companies
```
Comment (weedySeaDragon) [inline features/search_companies.feature:136]: "And" should be "Then"

File: features/user_account/member-home-account-page.feature [hunk 16]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
     And the user is paid through "2018-12-31"
 
 
+  @selenium @time_adjust
+  Scenario: Member can add subcategories to a category
+    Given I am logged in as "emma-member@example.com"
+    When I am on the "user account" page for "emma-member@example.com"
+    And I should see t("activerecord.models.business_category.other")
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:105]: "And" -> "Then"

File: features/user_account/member-home-account-page.feature [hunk 17]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
...
+    Then I click the first icon with CSS class "fa-edit"
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:106]: "Then" -> "When"

File: features/user_account/member-home-account-page.feature [hunk 18]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
...
+    And I select "light trim" in select list "subcategories-for-category-1"
+    And I select "custom cut" in select list "subcategories-for-category-1"
+    And I click on first t("save") button
+    And I should see "light trim"
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:110]: "And" -> "Then"

File: features/user_account/member-home-account-page.feature [hunk 19]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
...
+    And I should see "custom cut"
+    Then I click the first icon with CSS class "fa-edit"
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:112]: "Then" -> "When"

File: features/user_account/member-home-account-page.feature [hunk 20]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
...
+    And I should see "light trim"
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:113]: "And" -> "Then"

File: features/user_account/member-home-account-page.feature [hunk 21]
```diff
@@ -92,5 +98,24 @@ Feature:  Member home (account) page
...
+    And I should see "custom cut"
+    And I unselect "light trim" in select list "subcategories-for-category-1"
```
Comment (weedySeaDragon) [inline features/user_account/member-home-account-page.feature:115]: "And" -> "When"

File: spec/helpers/shf_applications_helper_spec.rb [hunk 22]
```diff
@@ -338,4 +338,39 @@
...
+    let(:app) do
+      app = build(:shf_application, num_categories: 0)
+      app.save(validate: false)
+      app.business_categories << cat
+      app.business_categories << cat.children
+      app
+    end
+
+    describe '#subcategories_list_in_parens' do
+
```
Comment (weedySeaDragon) [inline spec/helpers/shf_applications_helper_spec.rb:347]: need example for `!subcategories.present?`

File: app/helpers/shf_applications_helper.rb [hunk 23]
```diff
@@ -4,6 +4,24 @@ def can_edit_state?
     policy(@shf_application).permitted_attributes_for_edit.include? :state
   end
 
+  def business_categories_str(application)
+    cats_str = ''
+    application.business_categories.roots.order(:name).each do |category|
+      cats_str += ', ' unless cats_str.empty?
```
Comment (weedySeaDragon) [inline app/helpers/shf_applications_helper.rb:10]: what's the point of this line?  when would `cats_str` not be empty?  (If there is a case, then there needs to be an example/test for it in the `spec`)

File: app/models/business_category.rb [hunk 24]
```diff
@@ -1,8 +1,28 @@
...
 
-  validates_presence_of :name
+  PARENT_AND_CHILD_NAME_SEPARATOR = ' >> '
+
+  validates :name, presence: true
 
   has_and_belongs_to_many :shf_applications
   has_many :companies, through: :shf_applications
+
+  def self.for_search
```
Comment (weedySeaDragon) [inline app/models/business_category.rb]: Because `for` is a keyword and has a specific meaning, I'd suggest renaming this method.  (Also "search" is specific to how it is used, not what it does.)

 The spec says
```
  it 'returns all categories with subcategories adjacent to parent category'
```
...so the name should be something that communicates that.   (sibling categories and subcategories)

File: spec/models/business_category_spec.rb [hunk 25]
```diff
@@ -22,4 +38,28 @@
   describe 'Validations' do
     it { is_expected.to validate_presence_of :name }
   end
+
+  describe '.for_search' do
+
```
Comment (weedySeaDragon) [inline spec/models/business_category_spec.rb:43]: need to include cases with no sibling (adjacent) categories, no subcategories

File: app/models/business_category.rb [hunk 26]
```diff
@@ -1,8 +1,28 @@
...
+
+    roots.order(:name).each do |category|
+      categories << category
+      categories += category.children.order(:name)
+    end
+
+    categories
+  end
+
+  def search_name
```
Comment (weedySeaDragon) [inline app/models/business_category.rb]: Rename the method:  "search" is how it is used, not what it does.  (I can see this being used elsewhere to display the full name with ancestry.)
Perhaps `full_ancestry_name` ?

File: spec/models/business_category_spec.rb [hunk 27]
```diff
@@ -22,4 +38,28 @@
...
+      cat1
+      cat2
+      expected = [cat1] + cat1.children.order(:name) + [cat2] + cat2.children.order(:name)
+      expect(BusinessCategory.for_search).to eq expected
+    end
+  end
+
+  describe '#search_name' do
+
+    it 'returns category name as-is' do
```
Comment (weedySeaDragon) [inline spec/models/business_category_spec.rb:54]: I'd add " for a top-level category" [but optional]

File: spec/models/business_category_spec.rb [hunk 28]
```diff
@@ -22,4 +38,28 @@
...
+      expect(cat1.search_name).to eq cat1.name
+    end
+
+    it 'returns subcategory name with preface including category name' do
```
Comment (weedySeaDragon) [inline spec/models/business_category_spec.rb:58]: I'd reword this to "returns subcategory name prefaced with the parent category name, separated by >>"  [optional]

File: spec/models/company_spec.rb [hunk 29]
```diff
@@ -584,6 +584,10 @@
     let(:cat5) { create(:business_category, name: 'cat5') }
     let(:cat6) { create(:business_category, name: 'cat6') }
 
+    let(:cat1_subcat1) { cat1.children.create(name: 'cat1_subcat1') }
```
Comment (weedySeaDragon) [inline spec/models/company_spec.rb]: `categories_names` is not a scope.  It should really be moved out of this section.

File: app/models/company.rb [hunk 30]
```diff
@@ -164,11 +164,22 @@ def any_visible_addresses?
...
     # Fetch category names only for accepted applications from members
-    categories.select(:name).distinct.order(:name).joins(shf_applications: :user)
+    cats = categories.roots
+              .distinct
+              .order(:name)
+              .joins(shf_applications: :user)
               .where("users.member = ?", true)
-              .pluck(:name)
+
```
Comment (weedySeaDragon) [inline app/models/company.rb:174]: optional:  This could be simplified a wee bit by inserting this statement here:
```
    return cats.pluck(:name) unless include_subcategories
```
since there's no point in either getting any other attributes/columns if we're not going to include any subcategories.

(I don't think this will really have any performance benefit.)

File: spec/models/shf_application_spec.rb [hunk 31]
```diff
@@ -596,4 +596,37 @@
...
+
+    let(:app) do
+      app = build(:shf_application, num_categories: 0)
+      app.business_categories = [parent, child1, child2, child3]
+      app.save!
+      app
+    end
+
+    describe '#business_subcategories' do
+
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb:618]: need tests for:
```
  !business_category.is_root?
  !business_categories.include?(business_category)
```

File: spec/models/shf_application_spec.rb [hunk 32]
```diff
@@ -596,4 +596,37 @@
...
+      it 'returns all subcategories for a category associated with the app' do
+        expect(app.business_subcategories(parent)).to eq [child1, child2, child3]
+      end
+    end
+
+    describe '#set_business_subcategories' do
+
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb:633]: need tests for
```
!(business_category.is_root? &&
                      business_categories.include?(business_category))
```

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
Most (if not all) comments are small, style, or optional things.

File: app/models/company.rb (hunk 1 above)
Comment (patmbolger) [reply app/models/company.rb]: My intent was to have an exception raised if the method is called _without_ specifying that argument.  However, current uses of that method _all_ specify `false` for that value, so I went ahead and made that change.

File: app/models/shf_application.rb (hunk 2 above)
Comment (patmbolger) [reply app/models/shf_application.rb]: Done.

File: app/models/shf_application.rb (hunk 4 above)
Comment (patmbolger) [reply app/models/shf_application.rb]: Removed.

File: app/models/shf_application.rb (hunk 5 above)
Comment (patmbolger) [reply app/models/shf_application.rb]: That was due to an earlier version of the code calling this method.  Now removed.

File: spec/models/business_category_spec.rb (hunk 28 above)
Comment (patmbolger) [reply spec/models/business_category_spec.rb:58]: I want to avoid being too specific about what the preface contains (in case we drop the chevrons and go with, say, `<parent_cat_name>::<subcat_name>`

File: app/models/company.rb (hunk 30 above)
Comment (patmbolger) [reply app/models/company.rb:174]: Good suggestion.   That also simplifies the following code in the method.

File: features/conditions/company_incomplete_alerts.feature (hunk 10 above)
Comment (patmbolger) [reply features/conditions/company_incomplete_alerts.feature:62]: I changed this line in `step_definitions/membership_application_steps.rb` from:
```
ma.save(validate: (legacy_app ? false : true))
```
to:
```
ma.save!(validate: (legacy_app ? false : true))
```
That was masking a couple of bugs in our test data. This was one of them ("_" is not a valid character for an email domain, and our email validator was rejecting this attribute value).

File: app/helpers/shf_applications_helper.rb (hunk 23 above)
Comment (patmbolger) [reply app/helpers/shf_applications_helper.rb:10]: If the app has no associated categories then that var would otherwise be unassigned.

File: spec/models/business_category_spec.rb (hunk 28 above)
Comment (weedySeaDragon) [reply spec/models/business_category_spec.rb:58]: agreed.  then how about "... with the separator between them" or "... with a separator after the parent"

**Review by weedySeaDragon — APPROVED** [review]

File: config/locales/sv.yml [hunk 33]
```diff
@@ -39,6 +39,7 @@ sv:
   search: &search Sök
   close: Stänga
   lastupdated: &last_updated Senast uppdaterad
+  action: Verkan
```
Comment (thesuss) [inline config/locales/sv.yml]: What's the context? if only admin then it's fine. otherwise its a different word depending on context

File: config/locales/sv.yml [hunk 34]
```diff
@@ -264,6 +266,7 @@ sv:
         name: &business_category_name  Namn
         description: &description  Beskrivning
         subcategory: &subcategory Underkategori
+        subcategories: Kategorier
```
Comment (thesuss) [inline config/locales/sv.yml]: Underkategorier
(right?)

File: config/locales/sv.yml [hunk 35]
```diff
@@ -539,6 +542,15 @@ sv:
       files_delivery_method: Du har valt att leverera filer via
       method_chosen_when: (vald %{date})
 
+    business_category_display_row:
+      category_no_subcategories_html: Den här kategorin har ingen
+         underkategorier ännu. Du kan begära %{mailto_admin} att tilldela
```
Comment (thesuss) [inline config/locales/sv.yml]: Den här kategorin har inga
         underkategorier ännu. Du kan be %{mailto_admin} att lägga till
         en eller flera underkategorier.

**Review by thesuss — CHANGES_REQUESTED** [review]

File: config/locales/sv.yml (hunk 33 above)
Comment (patmbolger) [reply config/locales/sv.yml]: This is a column header for actions the user can take:

1. Associate themselves with one or more subcategories

2. Bring up an email form to send a request to the admin to add subcategory(s) to a category.

Comment (thesuss) [reply config/locales/sv.yml]: Does it need a header? 
-- No :)

File: config/locales/sv.yml (hunk 35 above)
Comment (patmbolger) [reply config/locales/sv.yml]: @thesuss What should "administrator" be (in Swedish)? Or, more precisely, how should we refer to the person/role used for the `mail_to` element?

Here is the code that defines `mailto_admin` in this translation:
```
mail_to_admin = mail_to(ENV['SHF_REPLY_TO_EMAIL'], 'administrator',
                                  subject: t('.subcategories_for',
                                             category_name: business_category.name))
```

I don't see anything in our locale files for "system administrator" (by itself), or "member administrator".

There is this reference to system administrator:

```
payments:
    create:
      page_title: &page_title '%{payment_type} betalning av: %{user_name}'
      title: *page_title
      something_wrong: 'Något gick fel - kontakta systemadministratören på %{admin_email}'
```

**Review by thesuss — APPROVED** [review]
